File: JSPTH/Math/Algebra.py
import math



class Algebra():
    pass



def quadratic_solver(a: float, b: float, c: float) -> tuple:
    """
    Solves the quadratic equation ax^2 + bx + c = 0 using the quadratic formula.
    
    Parameters:
    a (float): Coefficient of x^2.
    b (float): Coefficient of x.
    c (float): Constant term.
    
    Returns:
    tuple: Two solutions of the quadratic equation.
    
    Raises:
    ValueError: If the equation has no real solutions.
    """
    discriminant = b**2 - 4*a*c
    if discriminant < 0:
        raise ValueError("This equation has no real solutions.")
    
    sol1 = (-b + math.sqrt(discriminant)) / (2 * a)
    sol2 = (-b - math.sqrt(discriminant)) / (2 * a)
    
    return sol1, sol2



def linear_solver(a: float, b: float) -> float:
    """
    Solves the linear equation ax + b = 0.
    
    Parameters:
    a (float): Coefficient of x.
    b (float): Constant term.
    
    Returns:
    float: The solution of the linear equation.
    
    Raises:
    ValueError: If a is zero (which would not be a valid equation).
    """
    if a == 0:
        raise ValueError("The equation is invalid (a cannot be zero).")
    
    return -b / a


def exponential_solver(a: float, b: float) -> float:
    """
    Solves the exponential equation a^x = b.
    
    Parameters:
    a (float): The base.
    b (float): The result.
    
    Returns:
    float: The value of x.
    
    Raises:
    ValueError: If the equation is invalid.
    """
    if a <= 0 or b <= 0:
        raise ValueError("Base and result must be positive.")
    
    return math.log(b) / math.log(a)



def logarithm_solver(a: float, b: float) -> float:
    """
    Solves the logarithmic equation log_a(b).
    
    Parameters:
    a (float): The base of the logarithm.
    b (float): The result of the logarithm.
    
    Returns:
    float: The value of log_a(b).
    
    Raises:
    ValueError: If the base or result is not valid.
    """
    if a <= 0 or b <= 0 or a == 1:
        raise ValueError("Base and result must be positive and base cannot be 1.")
    
    return math.log(b) / math.log(a)

File: TypeScript/JSTS/Math/Algebra.ts
// File: MathUtils.ts

export class Algebra {
    // Quadratic Equation Solver: ax^2 + bx + c = 0
    static quadraticSolver(a: number, b: number, c: number): [number, number] | string {
        const discriminant = b ** 2 - 4 * a * c;
        if (discriminant < 0) {
            return "No real roots";
        }
        const root1 = (-b + Math.sqrt(discriminant)) / (2 * a);
        const root2 = (-b - Math.sqrt(discriminant)) / (2 * a);
        return [root1, root2];
    }

    // Linear Equation Solver: ax + b = 0
    static linearSolver(a: number, b: number): number {
        if (a === 0) {
            throw new Error("No solution (a = 0)");
        }
        return -b / a;
    }

    // Exponential Solver: a^x = b
    static exponentialSolver(a: number, b: number): number {
        if (a <= 0 || b <= 0) {
            throw new Error("Base and result must be positive");
        }
        return Math.log(b) / Math.log(a);
    }

    // Logarithm Solver: log_a(b)
    static logarithmSolver(a: number, b: number): number {
        if (a <= 0 || a === 1 || b <= 0) {
            throw new Error("Invalid base or input");
        }
        return Math.log(b) / Math.log(a);
    }

    // Polynomial Evaluation: Given coefficients [a_n, a_(n-1), ..., a_0], calculate P(x)
    static polynomialEval(coefficients: number[], x: number): number {
        return coefficients.reduce((result, coeff, index) => {
            return result + coeff * x ** (coefficients.length - index - 1);
        }, 0);
    }

    // Greatest Common Divisor (GCD) using Euclidean algorithm
    static gcd(a: number, b: number): number {
        while (b !== 0) {
            [a, b] = [b, a % b];
        }
        return Math.abs(a);
    }

    // Least Common Multiple (LCM)
    static lcm(a: number, b: number): number {
        return Math.abs(a * b) / Algebra.gcd(a, b);
    }

    // Binomial Coefficient: C(n, k)
    static binomialCoefficient(n: number, k: number): number {
        if (k < 0 || k > n) {
            return 0;
        }
        if (k === 0 || k === n) {
            return 1;
        }
        let result = 1;
        for (let i = 1; i <= k; i++) {
            result = (result * (n - i + 1)) / i;
        }
        return result;
    }

    // Arithmetic Progression: Sum of AP (a, a+d, a+2d, ..., a+(n-1)d)
    static arithmeticProgression(a: number, d: number, n: number): number {
        return (n / 2) * (2 * a + (n - 1) * d);
    }

    // Geometric Progression: Sum of GP (a, ar, ar^2, ..., ar^(n-1))
    static geometricProgression(a: number, r: number, n: number): number {
        if (r === 1) {
            return a * n;
        }
        return a * (1 - r ** n) / (1 - r);
    }

    // Geometric Series Sum: Infinite series (a, ar, ar^2, ...)
    static geometricSeriesSum(a: number, r: number): number {
        if (Math.abs(r) >= 1) {
            throw new Error("Infinite series does not converge for |r| >= 1");
        }
        return a / (1 - r);
    }
}
